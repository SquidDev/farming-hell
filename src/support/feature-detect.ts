export const isTouch = "ontouchstart" in window || navigator.maxTouchPoints > 0;
