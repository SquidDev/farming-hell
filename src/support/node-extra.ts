import { Agent as HttpsAgent, get as httpsGet } from "https";
import { ClientRequest, Agent as HttpAgent, IncomingMessage, get as httpGet } from "http";
import { promises as fs } from "fs";
import { StringDecoder } from "string_decoder";
import { spawn } from "child_process";

/** Check if a file exists. */
export const fileExists = async (path: string): Promise<boolean> => {
  try {
    await fs.access(path);
    return true;
  } catch {
    return false;
  }
};

const httpAgent = new HttpAgent({ keepAlive: true });
const httpsAgent = new HttpsAgent({ keepAlive: true });

const get = (url: string | URL, callback?: (res: IncomingMessage) => void): ClientRequest => {
  const isHttps = url instanceof URL ? url.protocol === "https" : url.startsWith("https://");
  return isHttps ? httpsGet(url, { agent: httpsAgent }, callback) : httpGet(url, { agent: httpAgent }, callback);
};

/** Download a remote URL and return it as a string. */
export const getAsString = (url: string): Promise<string> => new Promise((resolve, reject) => get(url, result => {
  if (result.statusCode !== 200) {
    reject(new Error(`Expected ${url} to return 200, got ${result.statusCode ?? "error"}`));
    return;
  }

  const builder = new StringDecoder("utf-8");
  let body = "";
  result.on("data", chunk => body += builder.write(chunk as Buffer));
  result.on("end", () => {
    body += builder.end();
    resolve(body);
  });
}).on("error", e => reject(e)));

/** Download a remote URL and save it to a file. This skips downloading if the file already exists, unless "force" is passed. */
export const getAsFile = async (url: string, path: string, force = false): Promise<void> => {
  if (await fileExists(path) && !force) return;

  console.log(`Downloading ${url} to ${path}`);
  await new Promise((resolve, reject) => {
    const doReject = async (e: unknown): Promise<void> => {
      await fs.unlink(path);
      reject(e);
    };

    get(url, result => {
      if (result.statusCode !== 200) {
        reject(new Error(`Expected ${url} to return 200, got ${result.statusCode ?? "error"}`));
        return;
      }

      fs.writeFile(path, result).then(resolve, doReject);
    }).on("error", e => void doReject(e));
  });
};

/** Run a command and wait for it to finish. */
export const runCommand = (program: string, ...args: Array<string>): Promise<void> => new Promise((resolve, reject) => {
  const process = spawn(program, args, {
    stdio: "ignore",
  });

  let done = false;
  process.on("error", e => {
    if (done) return;
    done = true;
    reject(e);
  });

  process.on("close", code => {
    if (done) return;
    done = true;

    if (code === 0) {
      resolve();
    } else {
      reject(`Process exited with ${code ?? "unknown error"}`);
    }
  });
});

export type Task = () => Promise<void>;
export type TaskQueue = Array<Task>;

/**
 * Run a list of tasks in parallel, with a limit on the maximum number of tasks run at once.
 */
export const runTasksInParallel = async (tasks: TaskQueue, workers = 8): Promise<void> => {
  let started = 0;

  const workerTasks = [];
  const worker = async (): Promise<void> => {
    while (true) {
      const task = tasks.pop();
      if (!task) return;

      const start = process.uptime();
      started++;
      await task();
      const finish = process.uptime();
      if (finish - start > 0.1) console.log(`Run ${started}/${started + tasks.length} tasks`);
    }
  };

  for (let i = 0; i < workers; i++) workerTasks.push(worker());
  await Promise.all(workerTasks);
};
