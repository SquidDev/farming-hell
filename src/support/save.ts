/**
 * Save a Blob to a file
 *
 * @param filename The name of the file to save.
 * @param blob The blob we're trying to save.
 */
export default (filename: string, blob: Blob): void => {
  // Somewhat inspired by https://github.com/eligrey/FileSaver.js/blob/master/src/FileSaver.js
  // Goodness knows how well this works on non-modern browsers.
  const element = document.createElement("a") ;
  const url = URL.createObjectURL(blob);

  element.download = filename;
  element.rel = "noopener";
  element.href = url;

  setTimeout(() => URL.revokeObjectURL(url), 60e3);
  setTimeout(() => {
    try {
      element.dispatchEvent(new MouseEvent("click"));
    } catch (e) {
      const mouseEvent = document.createEvent("MouseEvents");
      mouseEvent.initMouseEvent("click", true, true, window, 0, 0, 0, 80, 20, false, false, false, false, 0, null);
      element.dispatchEvent(mouseEvent);
    }
  }, 0);
};
