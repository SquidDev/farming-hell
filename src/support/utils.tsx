import type { FunctionComponent } from "react";

/**
 * Small helper-function for building class names.
 *
 * This is especially useful when some classes should be conditionally appled, as you can write:
 * {@code classNames(f(x) ? "some-class" : undefined)}
 */
export const classNames = (...classes: Array<string | undefined>): string => classes.filter(x => !!x).join(" ");

/** Wraps an <img> tag for a specific Atlas Academy icon, preconfiguring the correct width and height. */
export const Icon: FunctionComponent<{ className?: string, src: string, alt: string, title?: string }> = ({ className, src, alt, title }) =>
  <img className={className} alt={alt} src={src} title={title} width="128" height="128" />;


// I love Tailwind, but still haven't figured out the right way to do sane styling for links.
// Oh web dev, how I tire of you.
export const ExternalLink: FunctionComponent<{ href: string }> = ({ href, children }) =>
  <a href={href} target="_blank" className="font-bold text-blue-600 hover:text-blue-900 active:text-blue-900 focus:text-blue-900" rel="noreferrer">
    {children}
  </a>;


// I love Tailwind, but still haven't figured out the right way to do sane styling for links.
// Oh web dev, how I tire of you.
export const LinkButton: FunctionComponent<JSX.IntrinsicElements["button"]> = ({ children, ...props }) =>
  <button type="button" className="font-bold text-blue-600 hover:text-blue-900 active:text-blue-900 focus:text-blue-900" {...props}>
    {children}
  </button>;
