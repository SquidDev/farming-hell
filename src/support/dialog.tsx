import { Dialog, DialogBackdrop, DialogPanel } from "@headlessui/react";
import type { FunctionComponent, ReactNode } from "react";

import { classNames } from "./utils";

const WrappedDialog: FunctionComponent<{ isOpen: boolean, setOpen: (open: boolean) => void, className?: string, children: ReactNode }> = ({ isOpen, setOpen, className, children }) =>
  <Dialog
    as="div"
    className="fixed inset-0 z-10 overflow-y-auto transition duration-300 ease-out data-[closed]:opacity-0"
    open={isOpen}
    onClose={() => setOpen(false)}
    transition
  >
    <DialogBackdrop transition className="fixed inset-0 bg-opacity-80 bg-black" />
    <div className="relative inset-0 flex w-screen items-center justify-center p-4">
      <DialogPanel className={classNames(
        "inline-block w-full p-6 my-8 overflow-hidden text-left align-middle bg-white shadow-xl rounded-2xl",
        className,
      )}>
        {children}
      </DialogPanel>
    </div>
  </Dialog>;

export default WrappedDialog;

export const DialogTitle = Dialog.Title;

export const DialogButton: FunctionComponent<JSX.IntrinsicElements["button"]> =
  ({ className, children, ...attrs }) =>
    <button
      type="button"
      className={classNames(
        "inline-flex justify-center px-4 py-2 text-sm font-medium border border-transparent rounded-md",
        "focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2",
        className ?? "hover:bg-blue-200 text-blue-900 bg-blue-100 focus-visible:ring-blue-500"
      )}
      {...attrs}
    >
      {children}
    </button>;
