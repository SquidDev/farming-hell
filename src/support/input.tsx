import { Field, FieldAttributes, FormikProps, GenericFieldHTMLAttributes, useField } from "formik";
import Tooltip from "rc-tooltip";
import "rc-tooltip/assets/bootstrap.css";
import { FunctionComponent, ReactNode, forwardRef, useEffect, useRef } from "react";
import isEqual from "react-fast-compare";

import { classNames } from "./utils";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const RefField: FunctionComponent<FieldAttributes<any>> = forwardRef(function RefField(props, ref) {
  return <Field innerRef={ref} {...props} />;
});

/**
 * Custom text input control for Formik. This displays error validation messages
 * within a tooltip (which displays on focus) as well as any other custom
 * tooltip contents.
 */
export const Input: FunctionComponent<{
  name: string,
  children: ReactNode,
} & GenericFieldHTMLAttributes> = ({ name, children, className: classes, ...props }) => {
  const [_field, meta, _helpers] = useField<string>(name);

  return <Tooltip
    trigger='focus'
    overlay={() => <>
      {meta.error ? <p className="font-semibold text-red-600 pb-1">{meta.error}</p> : undefined}
      {children}
    </>}
    placement='bottom'
    // @ts-ignore: Types for Tooltip are incomplete.
    blurDelay={0}
  >
    <RefField
      name={name}
      className={classNames(
        "border focus:ring-2 focus:ring-indigo-500 hover:ring-2 hover:ring-indigo-400",
        classes,
        meta.error ? "border-red-600" : "border-gray-300"
      )} {...props} />
  </Tooltip>;
};

/**
 * This should be used within a Formik component to submit it whenever values
 * change, rather than the form is submitted.
 */
export const SubmitOnChange = <T,>({ formik }: { formik: FormikProps<T> }): null => {
  const lastValues = useRef(formik.values);

  useEffect(() => {
    if (!isEqual(lastValues.current, formik.values) && formik.isValid) {
      lastValues.current = formik.values;
      void formik.submitForm();
    }
  }, [
    lastValues,
    formik.values,
    formik.initialValues,
    formik,
  ]);

  return null;
};

export const Checkbox: FunctionComponent<GenericFieldHTMLAttributes> = ({ children, ...props }) =>
  <label className="flex items-center space-x-3">
    <Field
      type="checkbox"
      className="checked:form-tick appearance-none h-6 w-6 border border-gray-300 rounded-md checked:bg-blue-600 checked:border-transparent focus:outline-none"
      {...props} />
    <span className="text-gray-900 font-medium">{children}</span>
  </label>;
