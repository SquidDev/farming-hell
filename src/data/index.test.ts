import { promises as fs } from "fs";

import { fsStore } from "../test";
import { fileExists } from "../support/node-extra";

const makeLookup = <K, V>(objs: Array<V>, keyFn: (o: V) => K): Map<K, V> => {
  const out = new Map<K, V>();
  for (const obj of objs) out.set(keyFn(obj), obj);
  return out;
};

const normaliseServant = (x : string): string => x.toLowerCase().replace(/-/g, " ");

test("Has a basic list of items", async () => {
  const store = await fsStore();
  expect(store.items.filter(x => x.alwaysDisplay).map(x => `${x.id} => ${x.name}`)).toMatchSnapshot();
});

test("Includes webcrow IDs", async () => {
  const store = await fsStore();

  const servants = [...store.servants].sort((a, b) => a.webcrowId - b.webcrowId);

  const fgoContents = await fileExists("_build/data/fgo_sim.json")
    ? await fs.readFile("_build/data/fgo_sim.json", "utf-8")
    : await fs.readFile("data/fgo_sim.json", "utf-8");
  const fgoSimServants = JSON.parse(fgoContents) as Array<{ id: number, name: string }>;
  const fgoSimById = makeLookup(fgoSimServants, x => x.id);

  const different: Array<string> = [];
  for (const servant of servants) {
    const simServant = fgoSimById.get(servant.webcrowId)?.name ?? "";

    if (normaliseServant(simServant) === normaliseServant(servant.name)) continue;
    different.push(`${servant.webcrowId} => ours=${servant.id}:'${servant.name}' theirs='${simServant}'`);
  }

  expect(different).toMatchSnapshot();
});
