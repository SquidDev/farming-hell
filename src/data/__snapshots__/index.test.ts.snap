// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Has a basic list of items 1`] = `
[
  "1 => QP",
  "9770400 => Blaze of Wisdom",
  "6001 => Gem of Saber",
  "6002 => Gem of Archer",
  "6003 => Gem of Lancer",
  "6004 => Gem of Rider",
  "6005 => Gem of Caster",
  "6006 => Gem of Assassin",
  "6007 => Gem of Berserker",
  "6101 => Magic Gem of Saber",
  "6102 => Magic Gem of Archer",
  "6103 => Magic Gem of Lancer",
  "6104 => Magic Gem of Rider",
  "6105 => Magic Gem of Caster",
  "6106 => Magic Gem of Assassin",
  "6107 => Magic Gem of Berserker",
  "6201 => Secret Gem of Saber",
  "6202 => Secret Gem of Archer",
  "6203 => Secret Gem of Lancer",
  "6204 => Secret Gem of Rider",
  "6205 => Secret Gem of Caster",
  "6206 => Secret Gem of Assassin",
  "6207 => Secret Gem of Berserker",
  "6503 => Proof of Hero",
  "6516 => Evil Bone",
  "6512 => Dragon Fang",
  "6505 => Void's Dust",
  "6522 => Fool's Chain",
  "6527 => Deadly Poisonous Needle",
  "6530 => Mystic Spinal Fluid",
  "6533 => Stake of Wailing Night",
  "6534 => Mystic Gunpowder",
  "6549 => Small Bell of Absolution",
  "6551 => Twilight Ceremonial Blade",
  "6552 => Unforgettable Ashes",
  "6554 => Obsidian Edge",
  "6502 => Seed of Yggdrasil",
  "6508 => Ghost Lantern",
  "6515 => Octuplet Crystals",
  "6509 => Serpent Jewel",
  "6501 => Phoenix Feather",
  "6510 => Eternal Gear",
  "6511 => Forbidden Page",
  "6514 => Homunculus Baby",
  "6513 => Meteor Horseshoe",
  "6524 => Great Knight Medal",
  "6526 => Shell of Reminiscence",
  "6532 => Refined Magatama",
  "6535 => Eternal Ice",
  "6537 => Giant's Ring",
  "6536 => Aurora Steel",
  "6538 => Soundless Bell",
  "6541 => Arrowhead of Malice",
  "6543 => Crown of Silvery Light",
  "6545 => Divine Leyline Spiritron",
  "6547 => Ball of Rainbow Thread",
  "6550 => Scale of Fantasy",
  "6553 => Sunscale",
  "6507 => Claw of Chaos",
  "6517 => Heart of the Foreign God",
  "6506 => Dragon's Reverse Scale",
  "6518 => Spirit Root",
  "6519 => Warhorse's Young Horn",
  "6520 => Tearstone of Blood",
  "6521 => Black Beast Grease",
  "6523 => Lamp of Evil-Sealing",
  "6525 => Scarab of Wisdom",
  "6528 => Primordial Lanugo",
  "6529 => Cursed Beast Gallstone",
  "6531 => Mysterious Divine Wine",
  "6539 => Reactor Core of Dawn",
  "6540 => Tsukumo Mirror",
  "6542 => Egg of Truth",
  "6544 => Fragment of a Twinkling Star",
  "6546 => Fruit of Eternity",
  "6548 => Flaming Oni Lantern Plant",
  "6999 => Crystallized Lore",
  "7001 => Saber Piece",
  "7002 => Archer Piece",
  "7003 => Lancer Piece",
  "7004 => Rider Piece",
  "7005 => Caster Piece",
  "7006 => Assassin Piece",
  "7007 => Berserker Piece",
  "7101 => Saber Monument",
  "7102 => Archer Monument",
  "7103 => Lancer Monument",
  "7104 => Rider Monument",
  "7105 => Caster Monument",
  "7106 => Assassin Monument",
  "7107 => Berserker Monument",
  "7999 => Holy Grail",
]
`;

exports[`Includes webcrow IDs 1`] = `
[
  "8 => ours=101800:'Altera' theirs='Attila'",
  "10 => ours=102600:'Chevalier d'Eon' theirs='Le Chevalier d'Eon'",
  "14 => ours=200500:'Atalante' theirs='Atalanta'",
  "17 => ours=300100:'Cú Chulainn' theirs='Cu Chulainn'",
  "18 => ours=300500:'Elisabeth Báthory' theirs='Elizabeth Bathory'",
  "20 => ours=300700:'Cú Chulainn (Prototype)' theirs='Cu Chulainn (Prototype)'",
  "21 => ours=300900:'Leonidas I' theirs='Leonidas'",
  "24 => ours=400600:'Georgios' theirs='Saint George'",
  "30 => ours=401900:'Martha' theirs='Saint Martha'",
  "38 => ours=502100:'Cú Chulainn' theirs='Cu Chulainn (Caster)'",
  "40 => ours=600200:'Hassan of the Cursed Arm' theirs='Cursed Arm Hassan'",
  "44 => ours=601300:'Phantom of the Opera' theirs='The Phantom of the Opera'",
  "49 => ours=700300:'Lu Bu Fengxian' theirs='Lu Bu'",
  "61 => ours=502600:'Elisabeth Báthory (Halloween)' theirs='Elizabeth Bathory (Caster)'",
  "64 => ours=301600:'Hektor' theirs='Hector'",
  "70 => ours=301300:'Scáthach' theirs='Scathach'",
  "72 => ours=101400:'Fergus mac Róich' theirs='Fergus Mac Roich'",
  "78 => ours=301900:'Artoria Pendragon (Alter)' theirs='Artoria Pendragon (Alter Lancer)'",
  "79 => ours=501000:'Paracelsus von Hohenheim' theirs='Paracelsus Van Hohenheim'",
  "88 => ours=301200:'Brynhild' theirs='Brynhildr'",
  "95 => ours=201800:'Gilgamesh (Child)' theirs='Young Gil'",
  "96 => ours=1100200:'Edmond Dantès' theirs='Edmond Dantes'",
  "97 => ours=701400:'Florence Nightingale' theirs='Nightingale'",
  "98 => ours=702000:'Cú Chulainn (Alter)' theirs='Cu Chulainn (Alter)'",
  "99 => ours=401300:'Queen Medb' theirs='Medb'",
  "102 => ours=301700:'Li Shuwen' theirs='Li Shuwen (Lancer)'",
  "107 => ours=1100100:'Aŋra Mainiiu' theirs='Angra Mainyu'",
  "110 => ours=600300:'Hassan of the Hundred Personas' theirs='Hundred-Faced Hassan'",
  "111 => ours=502000:'Irisviel (Holy Grail)' theirs='Irisviel (Dress of Heaven)'",
  "113 => ours=502200:'Xuanzang Sanzang' theirs='Xuanzang'",
  "114 => ours=702300:'Minamoto-no-Raikou' theirs='Minamoto no Yorimitsu'",
  "117 => ours=601900:'Fuuma "Evil-wind" Kotarou' theirs='Fuuma Kotarou'",
  "118 => ours=401200:'Ozymandias' theirs='Ramesses II'",
  "121 => ours=102000:'Lancelot' theirs='Lancelot (Saber)'",
  "124 => ours=600900:'Hassan of the Serenity' theirs='Hassan of Serenity'",
  "133 => ours=602400:'Scáthach (Assassin)' theirs='Scathach (Assassin)'",
  "135 => ours=900400:'Martha (Ruler)' theirs='Saint Martha (Ruler)'",
  "138 => ours=101000:'Elisabeth Báthory (Brave)' theirs='Elizabeth Bathory (Brave)'",
  "141 => ours=302600:'Jeanne d'Arc Alter Santa Lily' theirs='Jeanne d'Arc (Alter) (Santa Lily)'",
  "145 => ours=501800:'Gilgamesh' theirs='Gilgamesh (Caster)'",
  "146 => ours=302700:'Medusa' theirs='Medusa (Lancer)'",
  "148 => ours=302800:'Jaguar Warrior' theirs='Jaguar Man'",
  "151 => ours=602500:'"First Hassan"' theirs='King Hassan'",
  "154 => ours=202300:'James Moriarty' theirs='Archer of Shinjuku'",
  "156 => ours=1100500:'Hessian Lobo' theirs='Avenger of Shinjuku'",
  "157 => ours=602600:'Yan Qing' theirs='Assassin of Shinjuku'",
  "158 => ours=102900:'Arthur Pendragon (Prototype)' theirs='Arthur Pendragon'",
  "161 => ours=1000200:'Meltryllis' theirs='Meltlilith'",
  "166 => ours=503000:'Scheherazade' theirs='Caster of Nightless City'",
  "167 => ours=602700:'Wu Zetian' theirs='Assassin of Nightless City'",
  "168 => ours=702700:'Penthesilea' theirs='Berserker of El Dorado'",
  "169 => ours=402500:'Christopher Columbus' theirs='Rider of Resistance'",
  "177 => ours=402700:'Artoria Pendragon Alter (Rider)' theirs='Artoria Pendragon (Rider Alter)'",
  "179 => ours=302900:'Minamoto-no-Raikou (Lancer)' theirs='Minamoto no Yorimitsu (Lancer)'",
  "180 => ours=303000:'Pārvatī' theirs='Parvati'",
  "181 => ours=202100:'Tomoe Gozen' theirs='Archer Inferno'",
  "182 => ours=602900:'Mochizuki Chiyome' theirs='Assassin Paraiso'",
  "183 => ours=301400:'Houzouin Inshun' theirs='Hozoin Inshun'",
  "184 => ours=103200:'Yagyu Tajima-no-kami Munenori' theirs='Yagyu Munenori'",
  "185 => ours=603000:'Katou "Black Kite" Danzo' theirs='Kato Danzo'",
  "187 => ours=1000500:'Mecha Eli-chan Mk.II' theirs='Mecha Eli-chan'",
  "188 => ours=503300:'Circe' theirs='Caster of Okeanos'",
  "190 => ours=503400:'Queen of Sheba' theirs='Caster of Midrash'",
  "197 => ours=703000:'Atalante (Alter)' theirs='Atalanta (Alter)'",
  "205 => ours=603300:'Okada Izo' theirs='Okada Izou'",
  "207 => ours=203100:'Napoleon' theirs='Napoleon Bonaparte'",
  "210 => ours=503900:'Scáthach-Skadi' theirs='Scathach-Skadi'",
  "213 => ours=303400:'Ibaraki-Douji (Lancer)' theirs='Ibaraki Doji (Lancer)'",
  "215 => ours=703100:'Jeanne d'Arc Alter (Berserker)' theirs='Jeanne d'Arc (Berserker Alter)'",
  "218 => ours=2500300:'Mysterious Heroine XX' theirs='Mysterious Heroine XX(Foreigner)'",
  "219 => ours=103500:'Diarmuid Ua Duibhne' theirs='Diarmuid Ua Duibhne (Saber)'",
  "221 => ours=504000:'Shuten-Douji (Caster)' theirs='Shuten Douji'",
  "223 => ours=103600:'Prince of Lan Ling' theirs='Lanling Wang'",
  "226 => ours=603500:'Yu Mei-ren' theirs='Yu Miaoyi'",
  "227 => ours=900600:'Qin Shi Huang' theirs='Shi Huang Di'",
  "239 => ours=2300300:'Jinako Carigiri' theirs='Great Statue God'",
  "240 => ours=103900:'Lakshmi Bai' theirs='Lakshmibai'",
  "243 => ours=203300:'Aśvatthāman' theirs='Asvatthaman'",
  "248 => ours=403500:'Leonardo da Vinci' theirs='Leonardo Da Vinci (Rider)'",
  "262 => ours=604000:'Okita Souji (Assassin)' theirs='Okita J Soji'",
  "266 => ours=203900:'Florence Nightingale (Santa)' theirs='Nightingale (Santa)'",
  "271 => ours=204000:'Sei Shounagon' theirs='Sei Shonagon'",
  "275 => ours=304200:'Romulus=Quirinus' theirs='Romulus-Quirinus'",
  "279 => ours=504500:'Artoria Caster' theirs='Artoria Pendragon (Caster)'",
  "280 => ours=2300400:'Sessyoin Kiara' theirs='Sesshoin Kiara (Moon Cancer)'",
  "281 => ours=204200:'Illyasviel von Einzbern' theirs='Illyasviel von Einzbern(Archer)'",
  "282 => ours=703800:'Brynhild' theirs='Brynhildr (Berserker)'",
  "283 => ours=304400:'Yu Mei-ren' theirs='Yu Miaoy (Lancer)'",
  "285 => ours=104500:'Tomoe Gozen' theirs='Tomoe Gozen (Saber)'",
  "286 => ours=404100:'Murasaki Shikibu' theirs='Murasaki Shikibu (Rider)'",
  "292 => ours=1001000:'Ashiya Douman' theirs='Ashiya Doman'",
  "294 => ours=104600:'Ibuki-Douji' theirs='Ibuki Doji'",
  "300 => ours=901100:'Amor (Caren)' theirs='Amor(Caren)'",
  "305 => ours=105000:'Barghest' theirs='Fairy Knight Gawain'",
  "306 => ours=204300:'Baobhan Sith' theirs='Fairy Knight Tristan'",
  "307 => ours=304800:'Melusine' theirs='Fairy Knight Lancelot'",
  "312 => ours=105100:'Okita Souji (Alter)' theirs='Okita Soji (Saber Alter)'",
  "314 => ours=504800:'Charlotte Corday' theirs='Charlotte Corday (Caster)'",
  "315 => ours=901200:'Leonardo da Vinci' theirs='Leonardo Da Vinci (Ruler)'",
  "316 => ours=1101100:'Kama' theirs='Kama (Avenger)'",
  "317 => ours=404500:'Caenis' theirs='Caenis (Rider)'",
  "318 => ours=704100:'Sei Shounagon' theirs='Sei Shonagon (Berserker)'",
  "321 => ours=404600:'Elisabeth Báthory (Cinderella)' theirs='Elizabeth Bathory (Cinderella)'",
  "324 => ours=305000:'Sakamoto Ryouma' theirs='Sakamoto Ryoma'",
  "326 => ours=404300:'Taigong Wang' theirs='Tai Gong Wang'",
  "330 => ours=1001100:'Manannán mac Lir (Bazett)' theirs='Manannan mac Lir (Bazett)'",
]
`;
